[GraalVM 24.0] RegisterForReflectionITCase.testLambdaCapturing from integration-tests/main fails with quarkus main branch

### Describe the bug

We see two new failures in the native integration tests (`main` suite) since today. Last pass was yesterday with GraalVM revision `651d8b7252cf81cb93b2fe33a14beabf628d7e18`. First failing revision we know of is `ba61de654bfcd2d92cfc0980872e6f33fce793d5`.

Failures are in these two tests:
```
io.quarkus.it.main.QuarkusTestCallbacksITCase.testCallbackContextIsNotFailed
io.quarkus.it.main.RegisterForReflectionITCase.testLambdaCapturing
```

They fail with:

```
Error:  Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.201 s <<< FAILURE! -- in io.quarkus.it.main.RegisterForReflectionITCase
Error:  io.quarkus.it.main.RegisterForReflectionITCase.testLambdaCapturing -- Time elapsed: 0.038 s <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Response body doesn't match expectation.
Expected: a string starting with "Comparator$$Lambda$"
  Actual: Comparator$$Lambda/0xc7facabea7d185476e9e1cd9e01e009cda293ae10

	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:73)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:108)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:57)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure.validate(ResponseSpecificationImpl.groovy:512)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure$validate$1.call(Unknown Source)
	at io.restassured.internal.ResponseSpecificationImpl.validateResponseIfRequired(ResponseSpecificationImpl.groovy:696)
	at io.restassured.internal.ResponseSpecificationImpl.this$2$validateResponseIfRequired(ResponseSpecificationImpl.groovy)
	at jdk.internal.reflect.GeneratedMethodAccessor146.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:198)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:62)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at io.restassured.internal.ResponseSpecificationImpl.body(ResponseSpecificationImpl.groovy:107)
	at io.restassured.internal.ValidatableResponseOptionsImpl.body(ValidatableResponseOptionsImpl.java:238)
	at io.quarkus.it.main.RegisterForReflectionITCase.testLambdaCapturing(RegisterForReflectionITCase.java:64)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

and

```
Error:  Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.037 s <<< FAILURE! -- in io.quarkus.it.main.QuarkusTestCallbacksITCase
Error:  io.quarkus.it.main.QuarkusTestCallbacksITCase.testCallbackContextIsNotFailed -- Time elapsed: 0.004 s <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <false> but was: <true>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertFalse.failNotFalse(AssertFalse.java:63)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:36)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:31)
	at org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:227)
	at io.quarkus.it.main.QuarkusTestCallbacksITCase.testCallbackContextIsNotFailed(QuarkusTestCallbacksITCase.java:61)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

See: https://github.com/graalvm/mandrel/actions/runs/6844550940/job/18609486281#step:12:657

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/builditem/ArtifactResultBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/builditem/ArtifactResultBuildItem.java
@@ -16,9 +16,9 @@ public final class ArtifactResultBuildItem extends MultiBuildItem {
 
     private final Path path;
     private final String type;
-    private final Map<String, Object> metadata;
+    private final Map<String, String> metadata;
 
-    public ArtifactResultBuildItem(Path path, String type, Map<String, Object> metadata) {
+    public ArtifactResultBuildItem(Path path, String type, Map<String, String> metadata) {
         this.path = path;
         this.type = type;
         this.metadata = metadata;
@@ -32,7 +32,7 @@ public final class ArtifactResultBuildItem extends MultiBuildItem {
         return type;
     }
 
-    public Map<String, Object> getMetadata() {
+    public Map<String, String> getMetadata() {
         return metadata;
     }
 }

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
@@ -176,7 +176,7 @@ public class JarResultBuildStep {
     ArtifactResultBuildItem jarOutput(JarBuildItem jarBuildItem) {
         if (jarBuildItem.getLibraryDir() != null) {
             return new ArtifactResultBuildItem(jarBuildItem.getPath(), PackageConfig.JAR,
-                    Collections.singletonMap("library-dir", jarBuildItem.getLibraryDir()));
+                    Collections.singletonMap("library-dir", jarBuildItem.getLibraryDir().toString()));
         } else {
             return new ArtifactResultBuildItem(jarBuildItem.getPath(), PackageConfig.JAR, Collections.emptyMap());
         }

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java
@@ -10,11 +10,9 @@ import java.nio.file.Paths;
 import java.nio.file.StandardCopyOption;
 import java.util.ArrayList;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Locale;
-import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Collectors;
@@ -89,12 +87,8 @@ public class NativeImageBuildStep {
     @BuildStep(onlyIf = NativeBuild.class)
     ArtifactResultBuildItem result(NativeImageBuildItem image) {
         NativeImageBuildItem.GraalVMVersion graalVMVersion = image.getGraalVMInfo();
-        Map<String, Object> graalVMInfoProps = new HashMap<>();
-        graalVMInfoProps.put("graalvm.version.full", graalVMVersion.getFullVersion());
-        graalVMInfoProps.put("graalvm.version.version", graalVMVersion.getVersion());
-        graalVMInfoProps.put("graalvm.version.javaVersion", "" + graalVMVersion.getJavaVersion());
-        graalVMInfoProps.put("graalvm.version.distribution", graalVMVersion.getDistribution());
-        return new ArtifactResultBuildItem(image.getPath(), PackageConfig.NATIVE, graalVMInfoProps);
+        return new ArtifactResultBuildItem(image.getPath(), PackageConfig.BuiltInType.NATIVE.getValue(),
+                graalVMVersion.toMap());
     }
 
     @BuildStep(onlyIf = NativeSourcesBuild.class)

--- a/core/deployment/src/main/java/io/quarkus/runner/bootstrap/AugmentActionImpl.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/bootstrap/AugmentActionImpl.java
@@ -228,12 +228,10 @@ public class AugmentActionImpl implements AugmentAction {
         if (lastResult.getPath() != null) {
             properties.put("path", outputTargetBuildItem.getOutputDirectory().relativize(lastResult.getPath()).toString());
         }
-        Map<String, Object> metadata = lastResult.getMetadata();
+        Map<String, String> metadata = lastResult.getMetadata();
         if (metadata != null) {
-            for (Map.Entry<String, Object> entry : metadata.entrySet()) {
-                if (entry.getValue() instanceof String) {
-                    properties.put("metadata." + entry.getKey(), entry.getValue());
-                }
+            for (Map.Entry<String, String> entry : metadata.entrySet()) {
+                properties.put("metadata." + entry.getKey(), entry.getValue());
             }
         }
         try (FileOutputStream fos = new FileOutputStream(quarkusArtifactMetadataPath.toFile())) {

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/ArtifactResult.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/ArtifactResult.java
@@ -7,9 +7,9 @@ public class ArtifactResult {
 
     private final Path path;
     private final String type;
-    private final Map<String, Object> metadata;
+    private final Map<String, String> metadata;
 
-    public ArtifactResult(Path path, String type, Map<String, Object> metadata) {
+    public ArtifactResult(Path path, String type, Map<String, String> metadata) {
         this.path = path;
         this.type = type;
         this.metadata = metadata;
@@ -23,7 +23,7 @@ public class ArtifactResult {
         return type;
     }
 
-    public Map<String, Object> getMetadata() {
+    public Map<String, String> getMetadata() {
         return metadata;
     }
 }

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/RegisterForReflectionITCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/RegisterForReflectionITCase.java
@@ -5,6 +5,9 @@ import static org.hamcrest.Matchers.startsWith;
 
 import org.junit.jupiter.api.Test;
 
+import io.quarkus.test.junit.DisableIfBuiltWithGraalVMNewerThan;
+import io.quarkus.test.junit.DisableIfBuiltWithGraalVMOlderThan;
+import io.quarkus.test.junit.GraalVMVersion;
 import io.quarkus.test.junit.QuarkusIntegrationTest;
 import io.restassured.RestAssured;
 
@@ -56,12 +59,19 @@ public class RegisterForReflectionITCase {
     }
 
     @Test
-    public void testLambdaCapturing() {
-        final String resourceLambda = BASE_PKG + ".ResourceLambda";
-
+    @DisableIfBuiltWithGraalVMNewerThan(GraalVMVersion.GRAALVM_23_1_0)
+    public void testLambdaCapturingPre24_0() {
         // Starting with GraalVM 22.1 support Lambda functions serialization
         // (see https://github.com/oracle/graal/issues/3756)
         RestAssured.given().when().get("/reflection/lambda").then().body(startsWith("Comparator$$Lambda$"));
+    }
+
+    @Test
+    @DisableIfBuiltWithGraalVMOlderThan(GraalVMVersion.GRAALVM_24_0_0)
+    public void testLambdaCapturingPost23_1() {
+        // Starting with GraalVM 24.0 lambda class names match the ones from HotSpot
+        // (see https://github.com/oracle/graal/pull/7775)
+        RestAssured.given().when().get("/reflection/lambda").then().body(startsWith("Comparator$$Lambda/"));
     }
 
     private void assertRegistration(String expected, String queryParam) {

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/DisableIfBuiltWithGraalVMNewerThanCondition.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/DisableIfBuiltWithGraalVMNewerThanCondition.java
@@ -36,7 +36,7 @@ public class DisableIfBuiltWithGraalVMNewerThanCondition implements ExecutionCon
         Properties quarkusArtifactProperties = readQuarkusArtifactProperties(context);
         try {
             GraalVM.Version version = GraalVM.Version
-                    .of(quarkusArtifactProperties.getProperty("metadata.graalvm.version.version").lines());
+                    .of(quarkusArtifactProperties.getProperty("metadata.graalvm.version.full").lines());
             int comparison = annotationValue.getVersion().compareTo(version);
             if (comparison < 0) {
                 return ConditionEvaluationResult.disabled("Native binary was built with GraalVM{version=" + version.toString()

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/DisableIfBuiltWithGraalVMOlderThanCondition.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/DisableIfBuiltWithGraalVMOlderThanCondition.java
@@ -36,7 +36,7 @@ public class DisableIfBuiltWithGraalVMOlderThanCondition implements ExecutionCon
         Properties quarkusArtifactProperties = readQuarkusArtifactProperties(context);
         try {
             GraalVM.Version version = GraalVM.Version
-                    .of(quarkusArtifactProperties.getProperty("metadata.graalvm.version.version").lines());
+                    .of(quarkusArtifactProperties.getProperty("metadata.graalvm.version.full").lines());
             int comparison = annotationValue.getVersion().compareTo(version);
             if (comparison > 0) {
                 return ConditionEvaluationResult.disabled("Native binary was built with GraalVM{version=" + version.toString()

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/GraalVMVersion.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/GraalVMVersion.java
@@ -3,7 +3,8 @@ package io.quarkus.test.junit;
 import io.quarkus.deployment.pkg.steps.GraalVM;
 
 public enum GraalVMVersion {
-    GRAALVM_23_1_0(GraalVM.Version.VERSION_23_1_0);
+    GRAALVM_23_1_0(GraalVM.Version.VERSION_23_1_0),
+    GRAALVM_24_0_0(GraalVM.Version.VERSION_24_0_0);
 
     private final GraalVM.Version version;
 

